calico_containers/pycalico/ipam.py

```diff
@@ -15,78 +15,472 @@
+        first time they request an auto-assigned IP.  For auto-assignment, a
+        host will allocate from a block it owns, or if all their currently
+        owned blocks get full, it will register itself as the owner of a new
+        block.  If all blocks are owned, and all the host's own blocks are
+        full, it will pick blocks at random until it can fulfil the request.
+        If you're really, really out of addresses, it will fail the request.
+
+        :param ip_version: 4 or 6, the IP version number.
+        :param num: Number of addresses to assign.
+        :param primary_key: allocation primary key for this request.
+        :param attributes: Contents of this dict will be stored with the
+        assignment and can be queried using get_assignment_attributes().  Must
+        be JSON serializable.
+        :param pool: (optional) if supplied, the pool to assign from,  If None,
+        automatically choose a pool.
+        :return:
+        """
 
-            candidate_address = self._get_next(pool, assigned_addresses)
-            if candidate_address is None:
-                # the pool is full, we can't allocate an address
-                return None
+        block_ids = iter(self._get_affine_blocks(my_hostname,
+                                                 ip_version,
+                                                 pool))
+        allocated_ips = []
+
+        num_remaining = num
+        while num_remaining > 0:
+            try:
+                block_id = block_ids.next()
+            except StopIteration:
+                break
+            ips = self._auto_assign_block(block_id,
+                                          num_remaining,
+                                          primary_key,
+                                          attributes)
+            allocated_ips.extend(ips)
+            num_remaining = num - len(allocated_ips)
+
+        # If there are still addresses to allocate, then we've run out of
+        # blocks with affinity.  Try to fullfil address request by allocating
+        # new blocks.
+        retries = RETRIES
+        while num_remaining > 0 and retries > 0:
+            retries -= 1
+            try:
+                new_block = self._new_affine_block(my_hostname,
+                                                   ip_version,
+                                                   pool)
+                # If successful, this creates the block and registers it to us.
+            except NoFreeBlocksError:
+                # No more blocks.
+                break
+            ips = self._auto_assign_block(new_block,
+                                          num_remaining,
+                                          primary_key,
+                                          attributes)
+            allocated_ips.extend(ips)
+            num_remaining = num - len(allocated_ips)
+        if retries == 0:  # pragma: no cover
+            raise RuntimeError("Hit Max Retries.")
```

Comment:
Yeah, all bets are off in this case.  The number of retries is meant to be so high that we would never hit it even under heavy load.  If multiple threads are trying to allocate from the same block one of them always wins the race, so the algorithm will make forward progress.

Hitting this error almost certainly means a bug in this code or etcd.